README.md
# Ljve Inspector

Forked from [node-inspector/node-inspector](https://github.com/node-inspector/node-inspector/).

## Overview

Ljve Inspector is a debugger interface for Node.js applications that uses the
Blink Developer Tools (formerly WebKit Web Inspector). It embeds extra features and
backported patches from node-inspector master branches and pull requests.

The original node-inspector project maintenance and support is sponsored by
[StrongLoop](http://strongloop.com).

### Table of Content

 * [Quick Start](#quick-start)
 * [Features](#features)
 * [Known Issues](#known-issues)
 * [Troubleshooting](#troubleshooting)
 * [Advanced Use](#advanced-use)
 * [Configuration](#configuration)
 * [Contributing](#contributing-code)
 * [Credits](#credits)

## Quick Start

#### Install

```sh
$ npm install -g node-inspector
```

#### Start

```sh
$ node-debug app.js
```

where ```app.js``` is the name of your main Node application JavaScript file.

See available configuration options [here](https://github.com/node-inspector/node-inspector#configuration)

#### Debug

The `node-debug` command will load Node Inspector in your default browser.

> **NOTE:** Node Inspector works in Chrome and Opera only. You have to re-open
the inspector page in one of those browsers if another browser
is your default web browser (e.g. Safari or Internet Explorer).

Node Inspector works almost exactly as the Chrome Developer Tools. Read the
excellent
[DevTools overview](https://developer.chrome.com/devtools/index)
to get started.

Other useful resources:
 - Documentation specific to Node Inspector provided by StrongLoop:
  [Debugging with Node Inspector](http://docs.strongloop.com/display/SLC/Debugging+applications)
 - Miroslav's talk
  [How to Debug Node Apps with Node Inspector](https://vimeo.com/77870960)
 - Danny's [screencasts](http://www.youtube.com/view_play_list?p=A5216AC29A41EFA8)
   (most likely outdated by now)
 - [Getting Started from scratch](http://github.com/node-inspector/node-inspector/wiki/Getting-Started---from-scratch)
   on wiki (most likely outdated by now)

## Features

The Blink DevTools debugger is a powerful JavaScript debugger interface.
Node Inspector supports almost all of the debugging features of DevTools, including:

* Navigate in your source files
* Set breakpoints (and specify trigger conditions)
* Step over, step in, step out, resume (continue)
* Inspect scopes, variables, object properties
* Hover your mouse over an expression in your source to display its value in
  a tooltip
* Edit variables and object properties
* Continue to location
* Break on exceptions
* Disable/enable all breakpoints
* CPU and HEAP profiling
* Network client requests inspection
* Console output inspection

### Cool stuff
* Node Inspector uses WebSockets, so no polling for breaks.
* Remote debugging
* [Live edit of running code](http://github.com/node-inspector/node-inspector/wiki/LiveEdit),
  optionally persisting changes back to the file-system.
* Set breakpoints in files that are not loaded into V8 yet - useful for
  debugging module loading/initialization.
* Embeddable in other applications - see [Embedding HOWTO](docs/embedding.md)
  for more details.

## Known Issues

* If there are symlink cycles then the [glob](https://github.com/isaacs/node-glob)
  module may take a long time to return results causing long delays at startup.
  The workaround is to disable preloading of scripts with `--no-preload`.
* Be careful about viewing the contents of Buffer objects,
  each byte is displayed as an individual array element;
  for most Buffers this will take too long to render.
* While not stopped at a breakpoint the console doesn't always
  behave as you might expect. See the
  [issue #146](https://github.com/node-inspector/node-inspector/issues/146).
* Break on uncaught exceptions does not work in all Node versions,
  you need at least v0.11.3 (see
  [node#5713](https://github.com/joyent/node/pull/5713)).
* Debugging multiple processes (e.g. cluster) is cumbersome.
  Read the following blog post for instructions:
  [Debugging Clustered Apps with Node-Inspector](http://strongloop.com/strongblog/whats-new-nodejs-v0-12-debugging-clusters/)

## Troubleshooting

#### My script runs too fast to attach the debugger.

The debugged process must be started with `--debug-brk`, this way the script is paused on the first line.

Note: `node-debug` adds this option for you by default.

#### I got the UI in a weird state.

When in doubt, refresh the page in browser

#### Can I debug remotely?

Yes. Node Inspector must be running on the same machine, but your browser can be anywhere.
Just make sure port 8080 is accessible.

#### How do I specify files to hide?

Create a JSON-encoded array.  You must escape quote characters when using a command-line option.

```sh
$ node-inspector --hidden='["node_modules/framework"]'
```

Note that the array items are interpreted as regular expressions.

#### UI doesn't load or doesn't work and refresh didn't help

Make sure that you have adblock disabled as well as any other content blocking scripts and plugins.

#### How can I (selectively) delete debug session metadata?

You may want to delete debug session metadata if for example Node Inspector gets in a bad state with some
watch variables that were function calls (possibly into some special c-bindings).  In such cases, even restarting
the application/debug session may not fix the problem.

Node Inspector stores debug session metadata in the HTML5 local storage.
You can inspect the contents of local storage and remove any items as
needed. In Google Chrome, you can execute any of the following in the JavaScript console:

```js
// Remove all
window.localStorage.clear()
// Or, to list keys so you can selectively remove them with removeItem()
window.localStorage
// Remove all the watch expressions
window.localStorage.removeItem('watchExpressions')
// Remove all the breakpoints
window.localStorage.removeItem('breakpoints')
```

When you are done cleaning up, hit refresh in the browser.

#### Node Inspector takes a long time to start up.

Try setting --no-preload to true. This option disables searching disk for *.js at startup.
Code will still be loaded into Node Inspector at runtime, as modules are required.

#### How do I debug Mocha unit-tests?

You have to start `_mocha` as the debugged process and make sure
the execution pauses on the first line. This way you have enough
time to set your breakpoints before the tests are run.

```sh
$ node-debug _mocha
```

#### How do I debug Gulp tasks?

If you are running on a Unix system you can simply run the following command.
The `$(which ..)` statement gets replaced with the full path to the gulp-cli.

```sh
$ node-debug $(which gulp) task
```

If you are running on Windows, you have to get the full path of `gulp.js`
to make an equivalent command:

```sh
> node-debug %appdata%\npm\node_modules\gulp\bin\gulp.js task
```
*You can omit the `task` part to run the `default` task.*

## Advanced Use

While running `node-debug` is a convenient way to start your debugging
session, there may come time when you need to tweak the default setup.

There are three steps needed to get you up and debugging:

#### 1. Start the Node Inspector server

```sh
$ node-inspector
```

You can leave the server running in background, it's possible to debug
multiple processes using the same server instance.

#### 2. Enable debug mode in your Node process

You can either start Node with a debug flag like:

```sh
$ node --debug your/node/program.js
```

or, to pause your script on the first line:

```sh
$ node --debug-brk your/short/node/script.js
```

Or you can enable debugging on a node that is already running by sending
it a signal:

1. Get the PID of the node process using your favorite method.
`pgrep` or `ps -ef` are good

    ```sh
    $ pgrep -l node
    2345 node your/node/server.js
    ```

2. Send it the USR1 signal

    ```sh
    $ kill -s USR1 2345
    ```

##### Windows

Windows does not support UNIX signals. To enable debugging, you can use
an undocumented API function `process._debugProcess(pid)`:

1. Get the PID of the node process using your favorite method, e.g.

    ```sh
    > tasklist /FI "IMAGENAME eq node.exe"

    Image Name                     PID Session Name        Session#    Mem Usage
    ========================= ======== ================ =========== ============
    node.exe                      3084 Console                    1     11,964 K
    ```

2. Call the API:

    ```sh
    > node -e "process._debugProcess(3084)"
    ```

#### 3. Load the debugger UI

Open http://127.0.0.1:8080/?ws=127.0.0.1:8080&port=5858 in the Chrome browser.

If you want to enable support for ECMAScript 6/7 through babeljs in console, open http://127.0.0.1:8080/?ws=127.0.0.1:8080&port=5858&babel=7 in the Chrome browser.

## Configuration

Both `node-inspector` and `node-debug` use [rc](https://npmjs.org/package/rc) module
to manage configuration options.

Places for configuration:
* command line arguments (parsed by [yargs](https://github.com/chevex/yargs))
* environment variables prefixed with ```node-inspector_```
* if you passed an option ```--config file``` then from that file
* a local ```.node-inspectorrc``` or the first found looking in ```./ ../ ../../
 ../../../``` etc.
* ```$HOME/.node-inspectorrc```
* ```$HOME/.node-inspector/config```
* ```$HOME/.config/node-inspector```
* ```$HOME/.config/node-inspector/config```
* ```/etc/node-inspectorrc```
* ```/etc/node-inspector/config```

All configuration sources that where found will be flattened into one object,
so that sources earlier in this list override later ones.

### Options

| Option | Alias | Default | Description |
| :------------------ | :-: | :-----: | :-------- |
| **general**
| --help              | -h  |         | Display information about available options.<br/>Use `--help -l` to display full usage info.<br/>Use `--help <option>` to display quick help on `option`.
| --version           | -v  |         | Display Node Inspector's version.
| --debug-port        | -d  | 5858    | Node/V8 debugger port.<br/>(`node --debug={port}`)
| --web-host          |     | 0.0.0.0 | Host to listen on for Node Inspector's web interface.<br/>`node-debug` listens on `127.0.0.1` by default.
| --web-port          | -p  | 8080    | Port to listen on for Node Inspector's web interface.
| **node-debug**
| --debug-brk         | -b  | true    | Break on the first line.<br/>(`node --debug-brk`)
| --nodejs            |     | []      | Pass NodeJS options to debugged process.<br/>(`node --option={value}`)
| --script            |     | []      | Pass options to debugged process.<br/>(`node app --option={value}`)
| --cli               | -c  | false   | CLI mode, do not open browser.
| **node-inspector**
| --save-live-edit    |     | false   | Save live edit changes to disk (update the edited files).
| --preload           |     | true    | Preload *.js files. You can disable this option<br/>to speed up the startup.
| --inject            |     | true    | Enable injection of debugger extensions into the debugged process. It's possible disable only part of injections using subkeys `--no-inject.network`. Allowed keys : `network`, `profiles`, `console`.
| --hidden            |     | []      | Array of files to hide from the UI,<br/>breakpoints in these files will be ignored.<br/>All paths are interpreted as regular expressions.
| --stack-trace-limit |     | 50      | Number of stack frames to show on a breakpoint.
| --ssl-key           |     |         | Path to file containing a valid SSL key.
| --ssl-cert          |     |         | Path to file containing a valid SSL certificate.

### Usage examples

#### Command line
##### Format
```
$ node-debug [general-options] [node-debug-options] [node-inspector-options] [script]
```

```
$ node-inspector [general-options] [node-inspector-options]
```
##### Usage

Display full usage info:
```
$ node-debug --help -l
```
Set debug port of debugging process to `5859`:
```
$ node-debug -p 5859 app
```
Pass `--web-host=127.0.0.2` to node-inspector. Start node-inspector to listen on `127.0.0.2`:
```
$ node-debug --web-host 127.0.0.2 app
```
Pass `--option=value` to debugging process:
```
$ node-debug app --option value
```
Start node-inspector to listen on HTTPS:
```
$ node-debug --ssl-key ./ssl/key.pem --ssl-cert ./ssl/cert.pem app
```
Ignore breakpoints in files stored in `node_modules` folder or ending in `.test.js`:
```
$ node-debug --hidden node_modules/ --hidden \.test\.js$ app
```
Add `--harmony` flag to the node process running the debugged script:
```
$ node-debug --nodejs --harmony app
```
Disable preloading of `.js` files:
```
$ node-debug --no-preload app
```

#### RC Configuration

Use dashed option names in RC files. Sample config file:
```js
{
  "web-port": 8088,
  "web-host": null,
  "debug-port": 5858,
  "save-live-edit": true,
  "preload": false,
  "hidden": ["\.test\.js$", "node_modules/"],
  "nodejs": ["--harmony"],
  "stack-trace-limit": 50,
  "ssl-key": "./ssl/key.pem",
  "ssl-cert": "./ssl/cert.pem"
}
```

## Contributing Code

Making Node Inspector the best debugger for node.js cannot be achieved without
the help of the community. The following resources should help you to get
started.

* [Contributing](https://github.com/node-inspector/node-inspector/wiki/Contributing)
* [Developer's Guide](https://github.com/node-inspector/node-inspector/wiki/Developer%27s-Guide)
* [Easy Picks](https://github.com/node-inspector/node-inspector/issues?direction=asc&labels=Easy+Pick&page=1&sort=updated&state=open)

## Credits

Maintainers

 - [Danny Coates](https://github.com/dannycoates) - the original author
   and a sole maintainer for several years.
 - [Miroslav Bajtoš](https://github.com/bajtos) - a current maintainer,
   sponsored by [StrongLoop](http://strongloop.com).
 - [3y3](https://github.com/3y3) - a current maintainer

Big thanks to the many contributors to the project, see [AUTHORS](AUTHORS).


ChangeLog.md
## 2015-06-29, Version 0.11.0

 * Update README.md (3y3)

 * front-end: Fix double sending of `togglePause` (3y3)

 * Fix `redirectToRoot` context (3y3)

 * Network: added new tests (3y3)

 * Network: fixed front-end (3y3)

 * Added Network tab (3y3)

 * Properly redirect to root (3y3)

 * Added /json/version http endpoint (Kenneth Auchenberg)

 * Added /json http endpoint (Kenneth Auchenberg)

 * Ignore errors during debugger closing. Fix tests. (3y3)

 * Fix inline sourcemaps path in node 0.10.x (Andres Suarez)

 * Send disconnect request on connection closing (3y3)


## 2015-06-15, Version 0.10.2

 * Fix test for iojs (3y3)

 * Avoid inline sourcemaps to be re-fetched from backend. (Marco Minetti)

 * Improve inline sourcemaps (3y3)

 * Check if source stored in app on source request (3y3)

 * Move `getScriptSourceById` to ScriptManager (3y3)

 * Console: fix max string length (3y3)

 * Console: fix caller script url (3y3)

 * Console: fix currentFrame searching (3y3)

 * Console: added special wrapper for assert func (3y3)

 * Fix index tests (3y3)

 * Fix url backward compatibility (3y3)

 * Fixed #667 (3y3)

 * Fix config web-host deprecation checker (3y3)

 * Added support for Web Notifications on debugger paused. (Marco Minetti)

 * Dropped /debug in favor of root URL. (Marco Minetti)


## 2015-05-13, Version 0.10.1

 * Bump dependencies (3y3)

 * Bump FrontendCommandHandler test timeout (3y3)

 * Try resolve script source on `debug` event (3y3)

 * More asynchronous Injector break handling (3y3)

 * Process `break` event in async mode (3y3)

 * Deprecate event caching in BreakEventHandler (3y3)

 * Use `wss:` WebSocket protocol if front-end is using SSL (Adam Biggs)

 * Small fix: redirection in debug-server (3y3)

 * README: remove traling whitespace (Miroslav Bajtoš)

 * ScriptFileStorage: follow symlinks, get realpaths (Jurgen Leschner)


## 2015-04-24, Version 0.10.0

 * Readme cleanup. (3y3)

 * Implemented "Save" for CPU and Heap profiles. (marco.minetti)

 * Fixed console test (3y3)

 * Fixed debug port not read from URI. (marco.minetti)

 * Fixed new watcher expression input. (Marco Minetti)

 * Fix console in debug-brk mode (3y3)

 * Add tests for BreakEventHandler (3y3)

 * Cache events from app (3y3)

 * Decrease time for FrontendCommandHandler test (3y3)

 * Fix tests for 0.12 (3y3)

 * Passing `session` object everywhere (3y3)

 * Fix console messageRepeatCounter (3y3)

 * Deprecate unuseful front-end changes (3y3)

 * Fix HeapProfilerAgent relative to new protocol.json (3y3)

 * Fix ProfilerAgent relative to new protocol.json (3y3)

 * Reload frontend on detach (3y3)

 * Allow to save modified files (3y3)

 * Disable unusable tabs (3y3)

 * Added PageAgent canEmulate and canScreencast callbacks (3y3)

 * Extend noop commands list (3y3)

 * Send page context to frontend (3y3)

 * Override watch expression (3y3)

 * Replace `(no domain)` string to `(core modules)` (3y3)

 * Replace worker title to '' (3y3)

 * Replace window title to `Node Inspector` (3y3)

 * Override WebInspector.UIString (3y3)

 * Open main script on startup (3y3)

 * Extend protocol by Console.showConsole (3y3)

 * Hide sources chrome specifics (3y3)

 * Remove unusable tabs from settings screen (3y3)

 * Unregister shortcuts (3y3)

 * Use default web socket to connect to inspector backend (3y3)

 * Update tools/generate-commands algorithm (3y3)

 * Frontend update: fetched from 2234 (3y3)

 * Disable debug conn. managment by Debugger Agent (3y3)

 * README: Add note regarding breaking bug (see: #534) (ELLIOTTCABLE)

 * Fixed console warning for profiler/node compatibility. (Marco Minetti)


## 2015-03-10, Version 0.9.2

 * Bump express min version to 4.8.0 (Alejandro Larre)

 * Documentation on contributing (jakub-g)


## 2015-02-24, Version 0.9.1

 * Move gitter badge to right place (3y3)

 * Added Gitter badge (The Gitter Badger)

 * README: fix typos (Wendy Leung)

 * Bump v8-debug to v0.4.2, v8-profiler to 5.2.4 (Jim Cummins)

 * .travis.yml: drop 0.11, add 0.12 and io.js (Miroslav Bajtoš)


## 2015-02-11, Version 0.9.0

 * tools: set +x for git-changelog.sh (Miroslav Bajtoš)

 * README note about preloading and symlinks (Jurgen Leschner)

 * use glob v4.3.5 - NOTE disables symlink following (Jurgen Leschner)

 * Enable latest Node v0.10 on Travis again (Miroslav Bajtoš)

 * Make sure to open URL in a webkit browser (jakub-g)

 * Fix: Breakpoints doesn't work for Windows (3y3)

 * lib: Fix typo (Seth Krasnianski)

 * Fix --hidden option for NODE_DEBUG mode (3y3)

 * travis: fix Node versions (Miroslav Bajtoš)


## 2015-01-12, Version 0.8.3

 * Fix release logic (3y3)


## 2015-01-12, Version 0.8.2

 * Fix release.sh for Windows (3y3)

 * Fix Console API getCallerFuncLocation (3y3)

 * Wrong name ScriptManager.normalizeV8Name (3y3)

 * Replace deprecated express sendfile with sendFile. (Timur Amirov)

 * front-end: fix non-expanding tree-element (tobiasviehweger)

 * Release script updated (3y3)

 * Check Windows case insensitive app name (3y3)

 * Avoid serialization of internal properties (3y3)


## 2014-12-08, Version 0.8.1

 * Update v8-profiler version (3y3)

 * Check NaN object in mirrors cache (3y3)

 * tools/git-changelog: always print the header (Miroslav Bajtoš)

 * tools/release: fixes and improvements (Miroslav Bajtoš)


## 2014-12-03, Version 0.8.0

 * Fixed issue #478 (jlu)

 * Modified DebuggerAgent.getFunctionDetails (3y3)

 * Added HeapProfiler._lookupHeapObjectId (3y3)

 * Console serializing and caching logic moved to InjectorServer (3y3)

 * README: fix command for debugging Gulp on Windows (thorn0)

 * Console API (3y3)

 * HeapProfiler API (3y3)

 * ChangeLog: update to the new markdown format (Miroslav Bajtoš)

 * Simplify the release process (Miroslav Bajtoš)

 * README: Fix broken link. (Marcelo Campos Rocha)

 * README: Add info on debugging Gulp tasks (Eric Eastwood)

 * Update README.MD (3y3)

 * Added usage examples (3y3)

 * New help system (3y3)


lib/config.js
var fs = require('fs'),
    rc = require('rc'),
    yargs = require('yargs'),
    path = require('path'),
    util = require('util');

module.exports = Config;

var definitions = {
  'help': {
    alias: 'h',
    description: 'Display information about avaible options.',
    usage: {
      '--help': '           display short list of avaible options',
      '--help <option>': '  display quick help on <option>',
      '--help -l': '        display full usage info'
    },
    _isNodeInspectorOption: true,
    _isNodeDebugOption: true,
    default: false
  },
  'version': {
    alias: 'v',
    type: 'boolean',
    description: 'Display Node Inspector\'s version.',
    usage: '--version',
    _isNodeInspectorOption: true,
    _isNodeDebugOption: true,
    default: false
  },
  'web-port': {
    alias: ['port', 'p'],
    type: 'number',
    description: 'Port to listen on for Node Inspector\'s web interface.',
    usage: {
      '--web-port 8081': '',
      '-p 8081': ''
    },
    _isNodeInspectorOption: true,
    default: 8080
  },
  'web-host': {
    type: 'string',
    description: 'Host to listen on for Node Inspector\'s web interface.',
    usage: {
      '--web-host 127.0.0.1': '',
      '--web-host www.example.com': ''
    },
    _isNodeInspectorOption: true,
    default: '0.0.0.0'
  },
  'debug-port': {
    alias: 'd',
    type: 'number',
    description: 'Node/V8 debugger port (`node --debug={port}`).',
    _isNodeInspectorOption: true,
    _isNodeDebugOption: true,
    default: 5858
  },
  'save-live-edit': {
    type: 'boolean',
    description: 'Save live edit changes to disk (update the edited files).',
    usage: {
      '--save-live-edit': '',
      '--no-save-live-edit': '    disable saving live edit changes to disk'
    },
    _isNodeInspectorOption: true,
    default: false
  },
  'preload': {
    type: 'boolean',
    description: 'Preload *.js files. You can disable this option to speed up the startup.',
    usage: {
      '--preload': '',
      '--no-preload': '    disable preloading *.js files'
    },
    _isNodeInspectorOption: true,
    default: true
  },
  'inject': {
    description: 'Enable/disable injection of debugger extensions into the debugged process.\n' +
               '  It`s posiible to disable only part of injections using subkeys.\n' +
               '  Available subkeys: network, profiles, console',
    usage: {
      '--inject': '',
      '--no-inject': '            disable injecting of debugger extensions',
      '--no-inject.network': '    disable injecting of debugger network extension only\n',
    },
    _isNodeInspectorOption: true,
    default: true
  },
  'plugins': {
    type: 'boolean',
    description: 'Enable plugin system.',
    usage: {
      '--plugins': '',
      '--no-plugins': '    disable plugin system'
    },
    _isNodeInspectorOption: true,
    default: true
  },
  'hidden': {
    type: 'string',
    description: 'Array of files to hide from the UI.' +
                ' Breakpoints in these files will be ignored.\n' +
               '  All paths are interpreted as regular expressions.',
    usage: {
      '--hidden .*\\.test\\.js$ --hidden node_modules/': 'ignore node_modules directory' +
      ' and all `*.test.js` files'
    },
    _isNodeInspectorOption: true,
    default: []
  },
  'stack-trace-limit': {
    type: 'number',
    description: 'Number of stack frames to show on a breakpoint.',
    _isNodeInspectorOption: true,
    default: 50
  },
  'ssl-key': {
    type: 'string',
    description: 'A file containing a valid SSL key.',
    usage: '--ssl-key ./ssl/key.pem --ssl-cert ./ssl/cert.pem',
    _isNodeInspectorOption: true,
    default: ''
  },
  'ssl-cert': {
    type: 'string',
    description: 'A file containing a valid SSL certificate.',
    usage: '--ssl-key ./ssl/key.pem --ssl-cert ./ssl/cert.pem',
    _isNodeInspectorOption: true,
    default: ''
  },
  'nodejs': {
    type: 'string',
    description: 'Pass NodeJS options to debugged process (`node --option={value}`).',
    usage: '--nodejs --harmony --nodejs --random_seed=2 app',
    _isNodeDebugOption: true,
    default: []
  },
  'debug-brk': {
    alias: 'b',
    type: 'boolean',
    description: 'Break on the first line (`node --debug-brk`).',
    _isNodeDebugOption: true,
    default: false
  },
  'cli': {
    alias: 'c',
    type: 'boolean',
    description: 'CLI mode, do not open browser.',
    usage: '--cli',
    _isNodeDebugOption: true,
    default: false
  }
};

function Config(argv, NODE_DEBUG_MODE) {
  var defaults = collectDefaultsFromDefinitions(NODE_DEBUG_MODE);
  var parsedArgv = parseArgs(argv);

  checkDeprecatedHiddenStyle(parsedArgv);

  var rcConfig = rc('node-inspector', defaults, parsedArgv);
  var config = normalizeOptions(rcConfig);

  checkDeprecatedNoPreloadStyle(config);
  checkDeprecatedWebHostStyle(config);

  util._extend(this, config);
}

function checkDeprecatedNoPreloadStyle(config) {
  if (config.noPreload !== undefined) {
    // Deprecated in v0.7.3
    console.warn('The config option `no-preload` is deprecated, use `preload` instead');
    config.preload = config.preload || !config.noPreload;
    delete config.noPreload;
  }
}

function checkDeprecatedWebHostStyle(config) {
  if (config.webHost === 'null' || config.webHost === null) {
    // Deprecated in v0.8.0
    console.warn('You use deprecated syntax for web-host option. Use 0.0.0.0 instead of null');
    config.webHost = '0.0.0.0';
  }
}

function checkDeprecatedHiddenStyle(parsedArgv) {
  if (parsedArgv.hidden.length == 1) {
    // Deprecated in v0.8.0
    var isDeprecatedDefinitionStyle = false,
        hidden;
    try {
      hidden = JSON.parse(parsedArgv.hidden[0]);
      isDeprecatedDefinitionStyle = util.isArray(hidden);
    } catch (e) {}

    if (isDeprecatedDefinitionStyle) {
      console.warn('You use deprecated syntax for hidden option.\n' +
                   'Use `--hidden value1 --hidden value2` instead');
      parsedArgv.hidden = hidden;
    }
  }
}

Config._collectDefaults = function(NODE_DEBUG_MODE) {
  var dashedKeyDefaults = collectDefaultsFromDefinitions(NODE_DEBUG_MODE);
  return normalizeOptions(dashedKeyDefaults);
};

Config.serializeOptions = function(options, filter) {
  filter = filter || {};
  var result = [];
  Object.keys(options).forEach(function(key) {
    if (filter[key]) return;

    var serializedOption = serializeOption(keyToDashedKey(key), options[key]);
    if (serializedOption !== '')
      result.unshift(serializedOption);
  });
  return result;
};

function serializeOption(key, value, parent) {
  parent = parent ? parent + '.' : '';
  var prefix = (key.length > 1 || parent ? '--' : '-') + parent;
  if (value === undefined) return '';
  if (value === true) {
    return prefix + key;
  } else if (value === false) {
    return prefix + key + '=false';
  } else if (util.isArray(value)) {
    if (!value.length) return '';
    return value.map(function(_value) {
      if (util.isRegExp(_value)) _value = _value.source;
      return prefix + key + '=' + _value;
    }).join(' ');
  } else if (value instanceof Object) {
    return Object.keys(value).map(function(_key) {
      return serializeOption(_key, value[_key], parent + key);
    }).join(' ');
  } else {
    return prefix + key + '=' + value;
  }
}

Config.filterOptions = function(filterExpression, options) {
  var filteredOptions = {};

  Object.keys(options)
    .filter(function(key) {
      var value = options[key];
      var dashedKey = keyToDashedKey(key);
      var definition = definitions[dashedKey];
      return filterExpression(key, value, definition);
    })
    .forEach(function(key) {
      filteredOptions[key] = options[key];
    });

  return filteredOptions;
};

Config.filterNodeDebugOptions = Config.filterOptions.bind(Config, function(key, value, definition) {
  return !definition || definition && definition._isNodeInspectorOption;
});

Config.filterDefaultValues = Config.filterOptions.bind(Config, function(key, value, definition) {
  return !definition || definition && definition.default !== value;
});

Config.prototype.showVersion = function() {
  console.log('Node Inspector v' + require('../package.json').version);
};

Config.prototype.showHelp = function(NODE_DEBUG_MODE) {
  var cmd = getCmd();

  var inspectorOptions = [];
  var nodeDebugOptions = [];
  Object.keys(definitions).forEach(function(key) {
    if (definitions[key]._isNodeDebugOption && NODE_DEBUG_MODE) {
      nodeDebugOptions.push(key);
    } else if (definitions[key]._isNodeInspectorOption) {
      inspectorOptions.push(key);
    }
  }, this);

  var inspectorPart = color('green', '[node-inspector-options]');
  var optionsPart = color('cyan', '[options]');
  var scriptPart = color('red', '[script [script-arguments]]');
  var configurationParts = [inspectorPart];
  if (NODE_DEBUG_MODE) {
    configurationParts.unshift(optionsPart);
    configurationParts.push(scriptPart);
  }

  if (typeof this.help == 'string') {
    //Display help for target option
    showOptionHelp(this.help);
  } else if (this.l) {
    //Display options format and options usage info
    console.log('Usage:\n    %s %s\n', cmd, configurationParts.join(' '));
    if (NODE_DEBUG_MODE) {
      nodeDebugOptions.forEach(showOptionHelp);
    }
    inspectorOptions.forEach(showOptionHelp);
  } else {
    //Display options format, options list and some help information
    console.log(
        'Usage:\n    %s %s\n', cmd, configurationParts.join(' '));

    if (NODE_DEBUG_MODE) {
      console.log(
        'Where %s is one or more of:\n' +
        '    %s\n', optionsPart, nodeDebugOptions.join('\n    '));
    }
    console.log(
        'Where %s is one or more of:\n' +
        '    %s\n', inspectorPart, inspectorOptions.join('\n    '));
    console.log('Use:' + formatUsage(definitions.help.usage) + '\n');

    if (NODE_DEBUG_MODE) {
      console.log(
        'The %s argument is resolved relative to the current working\n' +
        'directory. If no such file exists, then env.PATH is searched.\n',
        color('red', '[script]'));
      console.log(
        'The default mode is to break on the first line of the script, to run\n' +
        'immediately on start use `--no-debug-brk` or press the Resume button.\n');
      console.log(
        'When there is no script specified, the module in the current working\n' +
        'directory is loaded in the REPL session as `m`. This allows you to call\n' +
        'and debug arbitrary functions exported by the current module.\n');
    }

    console.log(
        'Configuration can be also stored in a \'.node-inspectorrc\' file,\n' +
        'see README for more details:\n' +
        '  https://github.com/node-inspector/node-inspector#configuration');
  }
};

function getCmd() {
  return process.env.CMD || path.basename(process.argv[1]);
}

function color(_color, string) {
  var colors = util.inspect.colors;
  return '\u001b[' + colors[_color][0] + 'm' + string +
         '\u001b[' + colors[_color][1] + 'm';
}

function showOptionHelp(option) {
  var info = definitions[option];

  if (info) {
    var optionLine = '--' + option;
    if (info.alias) {
      var aliases = [].concat(info.alias);
      optionLine += ', ' + aliases.map(function(alias) {
        var prefix = alias.length == 1 ? '-' : '--';
        return prefix + alias;
      }).join(', ');
    }
    console.log(color('green', optionLine));
    console.log('  ' + info.description);
    if (info.default !== undefined) {
      console.log('  Default: ' + color('yellow', JSON.stringify(info.default)));
    }
    if (info.usage) {
      var formattedUsage = formatUsage(info.usage);
      console.log('  Usage:' + formattedUsage);
    }
    console.log();
  } else {
    console.error('Description for %s not found', option);
  }
}

function formatUsage(usage) {
  var formattedUsage = '';
  var cmd = getCmd();
  if (typeof usage == 'object') {
    Object.keys(usage).forEach(function(key) {
      formattedUsage += '\n    ' + cmd + ' ' + key + ' ' + color('grey', usage[key]);
    });
  } else if (typeof usage == 'string') {
    formattedUsage += '\n    ' + cmd + ' ' + usage;
  }
  return formattedUsage;
}

function normalizeOptions(options) {
  var normalizedOptions = {};

  Object.keys(options).forEach(function(key) {
    var camelKey = keyToCamelKey(key);
    normalizedOptions[camelKey] = options[key];
  });

  checkHiddenOption(normalizedOptions);
  checkSslOptions(normalizedOptions);

  return normalizedOptions;
}

function checkHiddenOption(options) {
  function toRegExp(string) {
    return new RegExp(string, 'i');
  }
  options.hidden = [].concat(options.hidden || []).map(toRegExp);
}

function checkSslOptions(options) {
  function realPath(path) {
    if (!path) return '';

    path = fs.realpathSync(path);
    if (!fs.existsSync(path)) throw new Error('The file "' + path + '" does not exist');

    return path;
  }

  options.sslKey = realPath(options.sslKey);
  options.sslCert = realPath(options.sslCert);
}

function keyToCamelKey(key) {
  return key.replace(/-./g, function(letter) {
    return letter.slice(1).toUpperCase();
  });
}

function keyToDashedKey(key) {
  return key.replace(/[A-Z]/g, function(letter) {
    return '-' + letter.toLowerCase();
  });
}

function collectDefaultsFromDefinitions(NODE_DEBUG_MODE) {
  var options = {};

  Object.keys(definitions).forEach(function(key) {
    if (util.isArray(definitions[key].default)) {
      options[key] = definitions[key].default.slice();
    } else {
      options[key] = definitions[key].default;
    }
  });

  if (NODE_DEBUG_MODE) {
    options['web-host'] = '127.0.0.1';
    options['debug-brk'] = true;
  }

  return options;
}

function collectAliasesFromDefinitions() {
  var aliases = [];

  Object.keys(definitions).forEach(function(key) {
    if (definitions[key].alias)
      aliases = aliases.concat(definitions[key].alias);
  });

  return aliases;
}

function parseArgs(argv) {
  var argvParser = yargs.options(definitions);

  function wasCustomized(key) {
    var definition = definitions[key];
    if (!definition) return true;

    var options = [key].concat(definition.alias || []);

    return argv.some(function(arg) {
      var optionMatcher = /^(?:--no)?-+(.*?)(?:=.*)?$/;
      arg = (optionMatcher.exec(arg) || [])[1];
      if (arg && options.indexOf(arg) !== -1) return true;
    });
  }

  //Preparse --nodejs options
  var nodejsArgs = [];
  var nodejsIndex = argv.indexOf('--nodejs');
  while (nodejsIndex !== -1) {
    var nodejsArg = argv.splice(nodejsIndex, 2)[1];
    if (nodejsArg !== undefined) {
      nodejsArgs.push(nodejsArg);
    }
    nodejsIndex = argv.indexOf('--nodejs');
  }

  var options = argvParser.parse(argv);
  var script = options._[0];

  if (script) {
    // We want to pass along subarguments, but re-parse our arguments.
    argv = argv.slice(0, argv.indexOf(script) + 1);
    options = argvParser.parse(argv);
  }

  //filter options
  var aliases = collectAliasesFromDefinitions();
  Object.keys(options).forEach(function(key) {
    if (aliases.indexOf(key) > -1) {
      //Filter aliases
      delete options[key];
    } else if (options[key] instanceof Object) {
      //Ignore array options and object options
    } else if (/[A-Z]/.test(key)) {
      //filter camelKey options created by yargs
      delete options[key];
    } else if (!wasCustomized(key)) {
      //filter options with default values
      delete options[key];
    }
  });

  options['nodejs'] = nodejsArgs;

  return options;
}
